src/types.jl

```diff
@@ -17,25 +17,33 @@ struct ManagedSCSMatrix
     values::Vector{Cdouble}
     rowval::Vector{Int}
     colptr::Vector{Int}
-    m::Int
-    n::Int
-end
+    scsmatref::Base.RefValue{SCSMatrix}
 
-function ManagedSCSMatrix(m::Int, n::Int, A::SCSVecOrMatOrSparse)
-    A_sparse = sparse(A)
+    function ManagedSCSMatrix(m::Integer, n::Integer, A::SCSVecOrMatOrSparse)
+        A_sparse = sparse(A)
 
-    values = copy(A_sparse.nzval)
-    rowval = convert(Array{Int, 1}, A_sparse.rowval .- 1)
-    colptr = convert(Array{Int, 1}, A_sparse.colptr .- 1)
+        # we convert everything to make sure that no conversion will
+        # accidentally happen when ccalling `new`, so that `scsmatref`
+        # holds pointers to actual data stored in the fields.
 
-    return ManagedSCSMatrix(values, rowval, colptr, m, n)
-end
+        values = convert(Vector{Cdouble}, copy(A_sparse.nzval))
+        rowval = convert(Vector{Int}, A_sparse.rowval .- 1)
+        colptr = convert(Vector{Int}, A_sparse.colptr .- 1)
+
+        # scsmatref holds the reference to SCSMatrix created out of data in ManagedSCSMatrix.
+        # this way the reference to SCSMatrix always points to valid data
+        # as long as the ManagedSCSMatrix is not GC collected.
+        # One MUST
+        #   `Base.unsafe_convert(Ref{SCSMatrix}, scsmatref)`
```

Comment:
What's the reason for presenting two ways to do this instead of recommending the simpler way (specifying Ref in the call type)?

Reply:
one can use the ref in two different ways 
 * in `ccall` specifying the type of `Ref{SCSMatrix}` (it gets converted to c-compatible `Ptr{SCSMatrx}`), or 
 * by assign it to a `Ptr{SCSMatrix}` field in `SCSData`.

At the moment we do the latter, since `scs_init`, `scs_solve` and `scs_finish` and `ccalls` inside take `Ref{SCSData}`.